addons/spreadsheet_account/models/account.py

```diff
@@ -79,20 +72,185 @@ def _build_spreadsheet_formula_domain(self, formula_params):
+        for args in args_list:
+            codes = tuple({code for code in args["codes"] if code})
+            company_to_codes[args['company_id'] or self.env.company.id].update(codes)
+
+        default_domain = [('account_type', 'in', ['liability_payable', 'asset_receivable'])] if default_accounts else expression.FALSE_DOMAIN
+
+        all_accounts = dict()
+        for company_id, codes in company_to_codes.items():
+            code_domain = expression.OR(
+                [
+                    ('code', '=like', f'{code}%'),
+                ]
+                for code in codes
+            )
+            domain = expression.AND([
+                expression.OR([code_domain, default_domain]),
+                self.env['account.account']._check_company_domain(company_id),
+            ])
+            accounts = self.env['account.account'].with_company(company_id).search(domain)
+            all_accounts[company_id] = accounts
+
+        return all_accounts
+
+    def _get_all_lines(self, args_list, fields, timeline, accounts):
+        if not args_list or not fields:
+            return {}
+
+        all_lines = {}
+        aggregates = [field + ':sum' for field in fields]
+
+        include_initial_balance_account_ids = []
+        for company_id, acc in accounts.items():
+            include_initial_balance_account_ids += acc.with_company(company_id).filtered('include_initial_balance').ids
+
+        for period_num, period in enumerate(timeline):
+            if period_num == 0:
+                company_ids = list({args['company_id'] or self.env.company.id for args in args_list})
+                account_ids = include_initial_balance_account_ids
+            else:
+                args_list_in_period = [args for args in args_list if period in args['date_periods']]
+                company_ids = list({args['company_id'] or self.env.company.id for args in args_list_in_period})
+                subcodes = tuple({subcode for args in args_list_in_period for subcode in args['codes'] if subcode})
+                account_ids_in_period = self.env['account.account']
+                for company_id in company_ids:
+                    account_ids_in_period |= accounts[company_id].filtered(
+                        lambda account:
+                        account.with_company(company_id).code.startswith(subcodes)
+                        or account.with_company(company_id).include_initial_balance
+                    )
+                account_ids = account_ids_in_period.ids
+
+            domain = [
+                *self.env['account.move.line']._check_company_domain(company_ids),
+                ('account_id', 'in', account_ids),
+                ('parent_state', 'in', ('draft', 'posted')),
+                ('date', '>=', period[0]),
+                ('date', '<=', period[1]),
+            ]
+
+            lines_in_period = self.env['account.move.line'].with_context(allowed_company_ids=company_ids)._read_group(
```

Comment:
We actually don't need to do this `_read_group` for the first group (starting from 1900) if there are no accounts that need to include the initial balance.

And that read would actually be the most expensive one in a lot of cases :confused: